enhancements/node-healthcheck-operator.md

```diff
@@ -0,0 +1,435 @@
+
+
+```yaml
+apiVersion: remediation.medik8s.io/v1alpha1
+kind: NodeHealthCheck
+metadata:
+  name: example
+  # this CRD is cluster-scope, namespace is not needed  
+spec:
+  maxUnhealthy: "49%"
+  selector:
+    matchExpressions:
+      key: node-role.kubernetes.io/worker
+      operator: exists
+  unhealthyConditions:
+    - type: "Ready"
+      status:  "Unknown"
+      duration: "300s"
+   - type: "Ready"
+     status:  "False"
+     duration: "300s"
+  remediationTemplate:
+    kind: PoisonPillRemediationTemplate
+    apiVersion: poison-pill.medik8s.io/v1alphaX
+    name: M3_REMEDIATION_GROUP
+status:
+  healthyNodes: 5
+  obeservedNodes: 5
+```
+
+#### NodeHealthCheck controller
+Watch:
+- Watch NodeHealthCheck resources
+- Watch nodes with an event handler e.g controller runtime `EnqueueRequestsFromMapFunc` which returns NodeHealthCheck resources.
+
+Reconcile:
+- Periodically enqueue a reconcile request
+- Fetch all nodes targets. E.g:
+
+
+- Calculate the number of unhealthy targets.
+- Compare current number against `maxUnhealthy` threshold and temporary short
+ circuits if the threshold is met.
+- Create remedy objects for unhealthy targets according to the template in NHC object
+
+### Notes/Constraints
+
+The functionality of this controller have a lot in common with the Machine Health
+Check controller except that its creating objects and not
+performing any actions per se. When deploying this controller make
+sure 1. you have a controller implementing the API, e.g [poison pill]
+2. the machine health check controller is not running, or in case of a mixed
+ the node health checks selector avoid targeting nodes with machines.
+
+
+### Risks and Mitigations
+
+#### Remediation Interaction with ClusterAutoScalar
+
+Misconfiguration of nodes provisioning can cause nodes to fail to get operational.
+Remediation is blind to such problems and may interfere with diagnosing the problem.
+The backoff mechanism mentioned in MHC may be suited for Machines, because they
+have the MachineSet name that can be recorded and tracked. Without the machine API
+our knowledge of the nodes could not be handled in such a generic way. There is
+simply no generic way to record a node configuration origin in reliable way without
+putting provider specific annotations.
+Potentially the NHC resource could contain label selectors to track
+repeating remediation and mark them as [incrementally back-off](https://issues.redhat.com/browse/OCPCLOUD-800). e.g
+
+Suggested backoff specification under the NHC resource may look like this:
+```yaml
+kind: NodeHealthCheck
+...
+spec:
+  selector:
+     ...
+  # optional
+  backoff:
+    type: exponential
+    limit: 10m
+```
+
+Caveat of this approach is that we need to take care of conflicting or overlapping backoff definitions.
+Setting global catchall backoff strategy is impossible or just coarse without specifying a selector (how do we know what to backoff from?)
+
+Currently backoff is not implemented.
```

Comment:
Probably don't need this comment, enhancements are written before anything is implemented normally 😉